src/main/kotlin/com/safehill/kclient/network/remote/RemoteServer.kt

```diff
@@ -510,15 +526,69 @@ class RemoteServer(
         asset: EncryptedAsset,
         filterVersions: List<AssetQuality>,
     ) {
-        TODO("Not yet implemented")
+        val encryptedVersionByPresignedURL = emptyMap<String, EncryptedAssetVersion>().toMutableMap()
+
+        for (encryptedVersion in asset.encryptedVersions.values) {
+            if (!filterVersions.contains(encryptedVersion.quality)) {
+                continue
+            }
+
+            try {
+                val serverAssetVersion = serverAsset.versions.first {
+                    version -> version.versionName == encryptedVersion.quality.name
+                }
+                serverAssetVersion.presignedURL?.let {
+                    encryptedVersionByPresignedURL[it] = encryptedVersion
+                    S3Proxy.upload(encryptedVersion.encryptedData, it)
+                    this.markAsset(
+                        asset.globalIdentifier,
+                        encryptedVersion.quality,
+                        AssetDescriptorUploadState.Completed
+                    )
+                }
+            } catch (_: NoSuchElementException) {
+                println("no server asset provided for version ${encryptedVersion.quality.value}")
+                continue
+            }
+        }
+
+        val remoteServer = this
+        val coroutineScope = CoroutineScope(Job() + Dispatchers.IO)
```

Comment:
Could you explain why we are creating new coroutineScope rather than using [coroutineScope](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html)?

Reply:
What coroutinescope should we use in your opinion?

Reply:
The scope which inherits the job that is calling the upload function.